File: 16. 3Sum Closest/16. 3Sum Closest.py
class Solution:
    def threeSumClosest(self, nums: 'List[int]', target: 'int') -> 'int':
        total_nums = len(nums)
        if total_nums == 3:
            return sum(nums)
        else:
            solution = 0
            diff = 999999999
            nums.sort()
            
            for i in range(total_nums-2):
                j = i + 1
                k = total_nums - 1
                
                while j < k:
                    sums = nums[i] + nums[j] + nums[k]
                    
                    if sums == target:
                        return sums
                    else:
                        if abs(target - sums) < diff:
                            diff = abs(target - sums)
                            solution = sums
                        if sums < target:
                            j += 1
                        elif sums > target :
                            k -= 1
                    
            return solution

File: 16. 3Sum Closest/16. 3Sum Closest.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function(nums, target) {
    nums.sort((a, b) => a - b);
    let res = Infinity;
    
    for (let i = 0; i < nums.length - 2; i++) {
        if (i > 0 && nums[i] == nums[i-1]) {
            continue;
        }
        
        let j = i+1, k = nums.length - 1;
        while (j < k) {
            let total = nums[i] + nums[j] +  nums[k];
            if (Math.abs(target - total) < Math.abs(target - res)) {
                res = total;
            }
            
            if (total < target) {
                j++;
            }
            else if (total > target) {
                k--;
            }
            else {
                return res;
            }
        }
    }
    return res;
};
